### ISSUE TITLE
You cannot save a class that does not contain a property that represents the item ID

### ISSUE THREAD

**Issue by dthunziker** [mikeedwards83/Glass.Mapper#113]
@mikeedwards83 

I am getting the following error periodically in Page Editor. Using TDS codegen with the template shown below. In can see that IPromoProxy implements IGlassBase in the codegen file. An IIS reset fixes this for a short period, then it returns. SC 7.2 Update-3 (Glass.Mapper.Sc v3.2.3.49)

See also: http://stackoverflow.com/questions/27574397/sitecore-pageeditor-randomly-shows-error-with-glass-mapper

    You cannot save a class that does not contain a property that represents the item ID. Ensure that at least one property has been marked to contain the Sitecore ID. Type: Castle.Proxies.IPromoProxy

     at Glass.Mapper.Sc.Configuration.SitecoreTypeConfiguration.ResolveItem(Object target, Database database)
     at Glass.Mapper.Sc.GlassHtml.MakeEditable[T](Expression`1 field, Expression`1 standardOutput, T model, Object parameters, Context context, Database database, TextWriter writer)

Here is the template

    public partial interface IGlassBase{
		
		[SitecoreId]
		Guid Id{ get; }

		[SitecoreInfo(SitecoreInfoType.Language)]
        Language Language{ get; }

        [SitecoreInfo(SitecoreInfoType.Version)]
        int Version { get; }

		[SitecoreInfo(SitecoreInfoType.Url)]
        string Url { get; }
	}

	public abstract partial class GlassBase : IGlassBase{
		
		[SitecoreId]
		public virtual Guid Id{ get; private set;}

		[SitecoreInfo(SitecoreInfoType.Language)]
        public virtual Language Language{ get; private set; }

        [SitecoreInfo(SitecoreInfoType.Version)]
        public virtual int Version { get; private set; }

		[SitecoreInfo(SitecoreInfoType.Url)]
        public virtual string Url { get; private set; }
	}

**Comment by mikeedwards83** [comment]
Hi

Could you show me the model for IPromoProxy, could you also check to see if there are two IPromoProxy classes in your solutions?

How long does it take before the server shows this error? I am assuming that you can save the page sometimes but not other times.

Could you also share the code you are using to save data?

**Comment by dthunziker** [comment]
Hi Mike,

Below is the generated code for IPromo. I have multiple IPromo classes, however, they are under different namespaces. Could this be the issue?

I am now getting this error pretty consistently, and it appears in every <%= Editable(...) %> region on the page, regardless of the model. I am actually not saving any data, just attempting to open the page in Page Editor.

    namespace TxtAv.TDS.CodeGeneration.Models.sitecore.templates.Shared.Components
    {
 	/// <summary>
	/// IPromo Interface
	/// <para></para>
	/// <para>Path: /sitecore/templates/Shared/Components/Promo</para>	
	/// <para>ID: 743af4ab-68b0-45c4-ae9b-5d171a171604</para>	
	/// </summary>
	[SitecoreType(TemplateId=IPromoConstants.TemplateIdString )] //, Cachable = true
	public partial interface IPromo : IGlassBase , global::TxtAv.TDS.CodeGeneration.Models.sitecore.templates.Shared.Base.IBaseTemplate
	{
				        /// <summary>
					/// The Content field.
					/// <para></para>
					/// <para>Field Type: Rich Text</para>		
					/// <para>Field ID: 2c0d15cd-a5d3-4e99-90fc-039ef1ba4d80</para>
					/// <para>Custom Data: </para>
					/// </summary>
					[SitecoreField(IPromoConstants.ContentFieldName)]
					string Content  {get; set;}
			
					/// <summary>
					/// The Image field.
					/// <para></para>
					/// <para>Field Type: Image</para>		
					/// <para>Field ID: 01db481b-5ff3-4bc1-9237-cc3f1dea5fb4</para>
					/// <para>Custom Data: </para>
					/// </summary>
					[SitecoreField(IPromoConstants.ImageFieldName)]
					Image Image  {get; set;}
			
					/// <summary>
					/// The Link field.
					/// <para></para>
					/// <para>Field Type: General Link</para>		
					/// <para>Field ID: aee4c534-acd6-4c44-8f5e-dae9ab453480</para>
					/// <para>Custom Data: </para>
					/// </summary>
					[SitecoreField(IPromoConstants.LinkFieldName)]
					Link Link  {get; set;}
 
                                        ...

        }

**Comment by dthunziker** [comment]
I should note that I am working in a multi-site Sitecore instance.

When I use Page Editor on site 'A', everything is fine, however, if I load Page Editor in site 'B', I get these errors on both sites A and B moving forward. An IIS reset fixes it on site 'A', but 'B' always displays the errors.

**Comment by mikeedwards83** [comment]
I have an idea and I think it is a combination of those issue. I will do some more testing but I think it relates to line 252 of Context.cs and how Castle names it's proxy interfaces.

**Comment by dthunziker** [comment]
Thanks for your quick response!

That does look like it has something to do with this. If I understand correctly, although I have IPromo under different namespaces, type.Name could potentially be 'IPromo' for both.

I noticed that type.FullPath returns the full namespace plus interface name... is there any reason not to compare against that property instead?

**Comment by mikeedwards83** [comment]
Does one promo type inherit from another?

**Comment by dthunziker** [comment]
I do have an IBasePage for each Sitecore site. Each inherits from a shared/global IBasePage that's under a different namespace. I think this is more likely to be conflicting than IPromo, which I do have multiple ones of, but they don't inherit from one aother or a common IPromo.

**Comment by mikeedwards83** [comment]
Does each promo have an ID? I am really struggling to replicate the issue :-(

**Comment by dthunziker** [comment]
Yup, everything inherits from IGlassBase that I included in the initial post. I will do some renaming/re-factoring of my project to see if that fixes it.

**Comment by mikeedwards83** [comment]
Let me know if that works, at least it will narrow it and I can try some other test cases.

**Comment by dthunziker** [comment]
Thanks very much, Mike! I will report back with my findings.

**Comment by dthunziker** [comment]
I renamed all of my templates to be unique, however, this did not solve the problem. So, I shifted gears and began looking for suspicious code. The following use of a dynamic item turned out to be where this issue was originating. I have since replaced it and everything is fine with Page Editor now.

This was within a GlassUserControl<T> of type IBasePage, and the Guids being passed in were from Droptree fields on the Model.

    protected string GetUrl(Guid itemId)
    {
        dynamic item = SitecoreContext.GetItem<dynamic>(itemId);
        return item != null ? item.Url : String.Empty;
    }

**Comment by mikeedwards83** [comment]
Thanks, I will see if I can replicate.

**Comment by dthunziker** [comment]
So this issue returned again and now I'm not so sure where it's originating. As a test, I refactored all of my GlassUserControl<>, GlassCast<>, etc to use the concrete classes instead of interfaces, and this appears to be working so far.

**Comment by mikeedwards83** [comment]
I have added additional logging into the point where the failure occurs: 

https://github.com/mikeedwards83/Glass.Mapper/commit/7220f0fb6034edb6546e9eeeda1cee5c1f5a2a28

This will be on the build server at the moment (v3.3.0.157), this will allow us to get more information about what is causing the failure.

**Comment by dthunziker** [comment]
Thanks Mike, I will try this out and get back to you.

**Comment by mikeedwards83** [comment]
Did you get any more information?

**Comment by patrickbarrn** [comment]
I ran into the same issue that dthunziker outlined above using 8.0 rev. 150223 and Glass.Mapper.Sc 3.3.0.51. The issue seems to get resolved when I updated my code to use concrete classes instead of interfaces.

If it helps, here is the log output of the configuration dump: [GlassMapper - Failed item resolve log](https://gist.github.com/patrickbarrn/2e0f4b172401e897a855)

**Comment by mikeedwards83** [comment]
Could you send me the interface and all inherited interfaces?

**Comment by patrickbarrn** [comment]
Here is the TDS code generated classes and interfaces that the gist error logs refers to:
```
namespace ClientName.Global.Models.TemplateModels
{

	public partial interface IGlassBase
	{
		
		[SitecoreId]
		Guid Id{ get; }

		[SitecoreInfo(SitecoreInfoType.Language)]
	        Language Language{ get; }

	        [SitecoreInfo(SitecoreInfoType.Version)]
	        int Version { get; }

		[SitecoreInfo(SitecoreInfoType.Url)]
	        string Url { get; }

		[SitecoreInfo(SitecoreInfoType.DisplayName)]
	        string DisplayName { get; }

	        [SitecoreInfo(SitecoreInfoType.Name)]
	        string ItemName { get; }

	        [SitecoreParent(InferType = true)]
	        IGlassBase Parent { get; }

	        [SitecoreChildren(InferType = true)]
	        IEnumerable<IGlassBase> Children { get; }

	        [SitecoreInfo(SitecoreInfoType.TemplateId)]
	        Guid TemplateId { get; }
	}

	public abstract partial class GlassBase : IGlassBase
	{
		[SitecoreId]
		public virtual Guid Id{ get; set;}

		[SitecoreInfo(SitecoreInfoType.Language)]
	        public virtual Language Language{ get; private set; }

	        [SitecoreInfo(SitecoreInfoType.Version)]
	        public virtual int Version { get; private set; }

		[SitecoreInfo(SitecoreInfoType.Url)]
	        public virtual string Url { get; private set; }

		[SitecoreInfo(SitecoreInfoType.DisplayName)]
        	public virtual string DisplayName { get; protected set; }

	        [SitecoreInfo(SitecoreInfoType.Name)]
	        public virtual string ItemName { get; protected set; }

	        [SitecoreParent(InferType = true)]
	        public virtual IGlassBase Parent { get; protected set; }

	        [SitecoreChildren(InferType = true)]
	        public virtual IEnumerable<IGlassBase> Children { get; protected set; }

	        [SitecoreInfo(SitecoreInfoType.TemplateId)]
	        public Guid TemplateId { get; protected set; }
	}
}

namespace ClientName.Global.Models.TemplateModels.ProjectName.Flex_Items
{


 	/// <summary>
	/// IBottom_Flex_With_Image Interface
	/// <para></para>
	/// <para>Path: /sitecore/templates/User Defined/ProjectName/Flex Items/Bottom Flex with Image</para>	
	/// <para>ID: 17c73be6-f21b-4494-8d88-8cc12e185ae0</para>	
	/// </summary>
	[SitecoreType(TemplateId=IBottom_Flex_With_ImageConstants.TemplateIdString, AutoMap = true )] //, Cachable = true
	public partial interface IBottom_Flex_With_Image : IGlassBase 
	{
								/// <summary>
					/// The Heading field.
					/// <para></para>
					/// <para>Field Type: Single-Line Text</para>		
					/// <para>Field ID: eaefee35-003a-467b-a79a-fa4182109f6c</para>
					/// <para>Custom Data: </para>
					/// </summary>
					[SitecoreField(IBottom_Flex_With_ImageConstants.HeadingFieldName)]
					string Heading  {get; set;}
			
								/// <summary>
					/// The Image field.
					/// <para></para>
					/// <para>Field Type: Image</para>		
					/// <para>Field ID: 0d8d681b-1590-49ad-a304-002f048344f5</para>
					/// <para>Custom Data: </para>
					/// </summary>
					[SitecoreField(IBottom_Flex_With_ImageConstants.ImageFieldName)]
					Image Image  {get; set;}
			
								/// <summary>
					/// The Paragraph Text field.
					/// <para></para>
					/// <para>Field Type: Rich Text</para>		
					/// <para>Field ID: c05731b3-cf1a-47f8-9c26-66cbca5cd936</para>
					/// <para>Custom Data: </para>
					/// </summary>
					[SitecoreField(IBottom_Flex_With_ImageConstants.Paragraph_TextFieldName)]
					string Paragraph_Text  {get; set;}
			
				}


	public static partial class IBottom_Flex_With_ImageConstants{

			public const string TemplateIdString = "17c73be6-f21b-4494-8d88-8cc12e185ae0";
			public static readonly ID TemplateId = new ID(TemplateIdString);
			public const string TemplateName = "Bottom Flex with Image";

					
			public static readonly ID HeadingFieldId = new ID("eaefee35-003a-467b-a79a-fa4182109f6c");
			public const string HeadingFieldName = "Heading";
			
					
			public static readonly ID ImageFieldId = new ID("0d8d681b-1590-49ad-a304-002f048344f5");
			public const string ImageFieldName = "Image";
			
					
			public static readonly ID Paragraph_TextFieldId = new ID("c05731b3-cf1a-47f8-9c26-66cbca5cd936");
			public const string Paragraph_TextFieldName = "Paragraph Text";
			
			

	}

	
	/// <summary>
	/// Bottom_Flex_With_Image
	/// <para></para>
	/// <para>Path: /sitecore/templates/User Defined/ProjectName/Flex Items/Bottom Flex with Image</para>	
	/// <para>ID: 17c73be6-f21b-4494-8d88-8cc12e185ae0</para>	
	/// </summary>
	[SitecoreType(TemplateId=IBottom_Flex_With_ImageConstants.TemplateIdString, AutoMap = true)] //, Cachable = true
	public partial class Bottom_Flex_With_Image  : GlassBase, IBottom_Flex_With_Image 
	{
	   
						/// <summary>
				/// The Heading field.
				/// <para></para>
				/// <para>Field Type: Single-Line Text</para>		
				/// <para>Field ID: eaefee35-003a-467b-a79a-fa4182109f6c</para>
				/// <para>Custom Data: </para>
				/// </summary>
				[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Team Development for Sitecore - GlassItem.tt", "1.0")]
				[SitecoreField(IBottom_Flex_With_ImageConstants.HeadingFieldName)]
				public virtual string Heading  {get; set;}
					
						/// <summary>
				/// The Image field.
				/// <para></para>
				/// <para>Field Type: Image</para>		
				/// <para>Field ID: 0d8d681b-1590-49ad-a304-002f048344f5</para>
				/// <para>Custom Data: </para>
				/// </summary>
				[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Team Development for Sitecore - GlassItem.tt", "1.0")]
				[SitecoreField(IBottom_Flex_With_ImageConstants.ImageFieldName)]
				public virtual Image Image  {get; set;}
					
						/// <summary>
				/// The Paragraph Text field.
				/// <para></para>
				/// <para>Field Type: Rich Text</para>		
				/// <para>Field ID: c05731b3-cf1a-47f8-9c26-66cbca5cd936</para>
				/// <para>Custom Data: </para>
				/// </summary>
				[global::System.CodeDom.Compiler.GeneratedCodeAttribute("Team Development for Sitecore - GlassItem.tt", "1.0")]
				[SitecoreField(IBottom_Flex_With_ImageConstants.Paragraph_TextFieldName)]
				public virtual string Paragraph_Text  {get; set;}
					
			
	}
}
```

**Comment by mikeedwards83** [comment]
I am trying to reproduce this but can't are you able to give me any more information?

**Comment by patrickbarrn** [comment]
I downloaded the code from the 20150327 release, and started stepping through both the Glass.Mapper.Sc.Configuration.SitecoreTypeConfiguration & Glass.Mapper.Sc.GlassHtml classes. 

What I found was that on line 530 of the Glass.Mapper.Sc.GlassHtml class if I updated the line from:
<code>
var config = context.GetTypeConfiguration<SitecoreTypeConfiguration>(finalTarget);
</code>
To:
<code>
var config = context.GetTypeConfiguration<SitecoreTypeConfiguration>(finalTarget, false, false);
</code>

The exceptions I as seeing in the Experience Editor never show up. The IdConfig property on Glass.Mapper.Sc.Configuration.SitecoreTypeConfiguration is no longer null from the change above.

Not sure why the exception was showing up on this project or why clearing the app pool fixed the issue for a short time.

**Comment by dnandu** [comment]
Hi,

I am having same issue when trying to add a link. Is there a solution or workaround for this issue?

Thanks,
Dhara

**Comment by mikeedwards83** [comment]
Closing due to age of ticket please reopen if required.

**Comment by yannisgu** [comment]
Hi @mikeedwards83 

We encoured this problem too and could tackle down the source:

In a HttpRequst pipeline we had this piece of code: `glassContext.GetCurrentItem<object>(inferType: true)`. We supect that it was once called with a context item, where we have no matching type in the solution, so that once it was effectively resolved to a System.Object.

Starting then we got the error described in this issue. The problem is that the `Glass.Mapper.Context.TypeConfigurations` dictionary then contains an entry for System.Object. Then in `Glass.Mapper.Context.GetTypeConfigurationFromType` Glass Mapper tries to load the type configuration with the BaseType. In case the type is a proxy class, the base type is System.Object, so the System.Object TypeConfiguration was returned instead of the one of the interace.

Hope this help you to reproduce it and fix the issue!

**Comment by MinyieDiaz** [comment]
I am also getting the same issue periodically. An app pool recycle seems to fix it for a while. Going to try using concrete classes like some users here suggested. 

Using GlassMapper 3.2.3.50

**Comment by zbecknell** [comment]
Just wanted to share my experience with this error.  In my case, I was using a base interface that had the following properties:

```csharp
ID Id { get; set; }

Language Language { get; set; }

string Name { get; set; }

ID TemplateId { get; set; }

IEnumerable<ID> BaseTemplateIds { get; set; }

Item Item { get; set; }
```

I was relying mostly on auto-mapping throughout the project, but in a special case I needed to infer types.  For a while it all worked fine, but then I started getting the error.  My solution, which seems obvious now, was to be more explicit specifying my ID field with an attribute, like so:

```csharp
[SitecoreId]
ID Id { get; set; }
```

With the ID properly mapped, the error went away.
